clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::macros::{FormatArgsArg, FormatArgsExpn};
+use clippy_utils::{is_diag_trait_item, match_def_path, path_to_local_id, paths};
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind, HirId, Impl, ImplItem, ImplItemKind, Item, ItemKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, ExpnData, ExpnKind, Symbol};
+
+const FORMAT_MACRO_PATHS: &[&[&str]] = &[
+    &paths::FORMAT_ARGS_MACRO,
+    &paths::ASSERT_EQ_MACRO,
+    &paths::ASSERT_MACRO,
+    &paths::ASSERT_NE_MACRO,
+    &paths::EPRINT_MACRO,
+    &paths::EPRINTLN_MACRO,
+    &paths::PRINT_MACRO,
+    &paths::PRINTLN_MACRO,
+    &paths::WRITE_MACRO,
+    &paths::WRITELN_MACRO,
+];
```

Comment:
Done, but I'll leave making the changes to the FormatArgs lint to another PR so as not to make this huge.

How could we re-use the lint pass, when here we need to check we are inside the impl Display or Debug, etc. ?

Reply:
> Done, but I'll leave making the changes to the FormatArgs lint to another PR so as not to make this huge.

I still would like to avoid duplicating the list in this PR. I small refactor can be done in the other lint.

> How could we re-use the lint pass, when here we need to check we are inside the impl Display or Debug, etc. ?

It seems like it should be possible to merge the two lint passes and maintain all cases. But I may be underestimating the difficulty. Again I'll leave it up to you to decide if that ultimately makes sense to do.

Reply:
@jamesmcm did you decide against merging the lint passes or still looking into it?

Reply:
I took a look at it, in theory it should be possible, it's not trivial though since the format_args case also needs to work when the macro isn't nested. It might be best to do in another PR though.

Btw is there any way to share the pass without combining the source files (and error files)?

Reply:
If we combine them, I think it would be better to do it in this PR since it would be less churn and reviewing. The tests can still be separated per lint.

Reply:
Cool, looking at it in detail I don't think it is worth combining them. Since handling the nested macro difference means the code isn't really shared between the non format trait impl and format trait impl cases, so it's more just moving it around and I don't think that's worth merging them (since at that point you might as well merge #8253 too for example).

Reply:
@camsteffen considering this, I think this is mergeable as is?

Reply:
Sorry for the delay.

> (since at that point you might as well merge #8253 too for example).

I am planning to have that lint merged with this one at least. But if that doesn't change your decision, yes this is ready to merge!